File: python/count_solutions.py
def count_solutions(a, b):
    dp = [0] * (b + 1)
    dp[0] = 1

    for i in range(len(a)):
        for j in range(a[i], b + 1):
            dp[j] += dp[j - a[i]]
    return dp[b]

print(5 == count_solutions([1, 2, 3], 5))

File: ruby/count_solutions.rb
def count_solutions(a, b)
  dp = [0] * (b + 1)
  dp[0] = 1

  0.upto(a.length-1) do |i|
    a[i].upto(b) do |j|
      dp[j] += dp[j - a[i]]
    end
  end

  dp[b]
end

puts 5 == count_solutions([1,2,3], 5)
